                        Travail réalisé par Mame Anta SARR et Ayaba Tabitha Favour GAGLOZOU

############################################################################################################################

                                        Milestone 0 — Cadrage du projet

Ce projet porte sur l'analyse de données d'annonces Airbnb à Barcelone.
L'objectif est de construire progressivement des modèles de Machine Learning permettant
d'étudier et de prédire différentes caractéristiques des annonces, tout en respectant
une démarche rigoureuse de préparation, validation et interprétation des données.

In [1]:
import sys
from pathlib import Path

ROOT = Path.cwd()

while not (ROOT / "src").is_dir() and ROOT != ROOT.parent:
    ROOT = ROOT.parent

sys.path.insert(0, str(ROOT))

import numpy as np
import pandas as pd
from src.data import PATHS, SEED, describe_environment, load_raw, set_seed

pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 140)

print(f"seed = {set_seed(SEED)}")
print(describe_environment().to_string(index=False))

seed = 42
package version
 python  3.14.2
  numpy   2.5.2
 pandas   3.0.5
sklearn   1.9.0


                                            Vérification de l'environnement

L'environnement d'exécution est correctement initialisé avec une graine aléatoire
fixée à `42`, ce qui permet d'obtenir des résultats reproductibles lors des
opérations faisant intervenir de l'aléatoire.

L'environnement utilise actuellement Python `3.14.2`, NumPy `2.5.2`, Pandas `3.0.5`
et Scikit-learn `1.9.0`.

La vérification de ces versions permet de documenter l'environnement logiciel
utilisé pour le projet et de faciliter la reproductibilité des expériences.

Le train/test split est réalisé tôt dans le workflow afin d’éviter la fuite d’information entre les données d’entraînement 
et les données de test (data leakage). Les décisions de nettoyage, de transformation et de préparation des données doivent être apprises uniquement à partir du jeu d’entraînement, puis appliquées au jeu de test. Cela permet d’obtenir une évaluation plus réaliste des performances du modèle sur de nouvelles données.

In [2]:
import pandas as pd
import numpy as np

# Chargement du jeu de données
df = load_raw()

print(f"Nombre de lignes : {df.shape[0]}")
print(f"Nombre de colonnes : {df.shape[1]}")

Nombre de lignes : 15293
Nombre de colonnes : 90


In [3]:
df.head()

,id,listing_url,scrape_id,last_scraped,source,name,description,neighborhood_overview,picture_url,host_id,host_url,host_profile_id,host_profile_url,host_name,host_since,hosts_time_as_user_years,hosts_time_as_user_months,hosts_time_as_host_years,hosts_time_as_host_months,host_location,host_about,host_response_time,host_response_rate,host_acceptance_rate,host_is_superhost,...,availability_365,calendar_last_scraped,number_of_reviews,number_of_reviews_ltm,number_of_reviews_l30d,availability_eoy,number_of_reviews_ly,estimated_occupancy_l365d,estimated_revenue_l365d,first_review,last_review,review_scores_rating,review_scores_accuracy,review_scores_cleanliness,review_scores_checkin,review_scores_communication,review_scores_location,review_scores_value,license,instant_bookable,calculated_host_listings_count,calculated_host_listings_count_entire_homes,calculated_host_listings_count_private_rooms,calculated_host_listings_count_shared_rooms,reviews_per_month
0,18674,https://www.airbnb.com/rooms/18674,20260624162917,2026-06-25,city scrape,Huge flat for 8 people close to Sagrada Familia,110m2 apartment to rent in Barcelona. Located ...,NaN,https://a0.muscache.com/pictures/hosting/Hosti...,71615,https://www.airbnb.com/users/show/71615,1.462508e+18,https://www.airbnb.com/users/profile/146250835...,Maria Y Mireia Barcelona4Seasons,NaN,16.0,5.0,15.0,5.0,"Barcelona, Spain","We are Mireia and Maria, two multilingual entr...",NaN,NaN,NaN,f,...,123,2026-06-25,59,9,1,121,9,54,22086.0,2013-05-27,2026-05-28,4.41,4.47,4.59,4.66,4.64,4.83,4.38,Spain - National registration number<br />ESFC...,NaN,20,20,0,0,0.37
1,23197,https://www.airbnb.com/rooms/23197,20260624162917,2026-06-25,city scrape,CCIB Forum· Large Balcony · 5 min walk to CCIB,"Beautiful, spacious apartment with large balco...",NaN,https://a0.muscache.com/pictures/hosting/Hosti...,90417,https://www.airbnb.com/users/show/90417,1.462509e+18,https://www.airbnb.com/users/profile/146250935...,Marnie,NaN,16.0,3.0,15.0,5.0,"Catalonia, Spain","Hi there,\n\nI’m marnie, originally from Austr...",NaN,NaN,NaN,t,...,239,2026-06-25,99,11,1,159,11,66,25608.0,2011-03-15,2026-06-07,4.84,4.95,4.91,4.94,4.99,4.69,4.71,Spain - National registration number<br />ESFC...,NaN,1,1,0,0,0.53
2,34981,https://www.airbnb.com/rooms/34981,20260624162917,2026-06-25,city scrape,VIDRE HOME PLAZA REAL on LAS RAMBLAS,Spacious apartment for large families or group...,NaN,https://a0.muscache.com/pictures/c4d1723c-e479...,73163,https://www.airbnb.com/users/show/73163,1.462508e+18,https://www.airbnb.com/users/profile/146250844...,Andres,NaN,16.0,5.0,15.0,5.0,"Barcelona, Spain","Hello I am a Professional designer, a traveler...",NaN,NaN,NaN,t,...,329,2026-06-25,304,37,1,159,33,222,88163.0,2010-10-03,2026-05-26,4.59,4.64,4.68,4.72,4.75,4.65,4.48,Spain - National registration number<br />ESFC...,NaN,2,2,0,0,1.59
3,36763,https://www.airbnb.com/rooms/36763,20260624162917,2026-07-03,previous scrape,In front of the beach,NaN,NaN,https://a0.muscache.com/pictures/airflow/Hosti...,158596,https://www.airbnb.com/users/show/158596,1.462511e+18,https://www.airbnb.com/users/profile/146251123...,Ester,NaN,15.0,11.0,15.0,6.0,"Barcelona, Spain","I live in a neighborhood ""Barceloneta"" next to...",NaN,NaN,NaN,f,...,38,2026-07-03,108,0,0,38,0,0,0.0,2011-09-28,2023-12-13,4.79,4.88,4.71,4.93,4.92,4.92,4.82,NaN,NaN,1,0,1,0,0.60
4,40983,https://www.airbnb.com/rooms/40983,20260624162917,2026-06-25,city scrape,Architect's Loft near Gaudí & Passeig de Gràcia,Stay in a charming architect-designed loft in ...,NaN,https://a0.muscache.com/pictures/miso/Hosting-...,177617,https://www.airbnb.com/users/show/177617,1.462512e+18,https://www.airbnb.com/users/profile/146251182...,Joaquin,NaN,15.0,11.0,15.0,5.0,"Barcelona, Spain","Hello\nI'm Joaquín, from Barcelona. I'm Archit...",NaN,NaN,NaN,t,...,289,2026-06-25,450,56,7,114,64,255,57663.0,2011-06-16,2026-06-19,4.56,4.62,4.59,4.85,4.86,4.91,4.52,Barcelona - Regional registration number<br />...,NaN,5,2,3,0,2.46


                                    Première observation du jeu de données

L'affichage des cinq premières observations permet d'identifier la structure
générale du jeu de données. Chaque ligne correspond à une annonce Airbnb et
les colonnes décrivent différentes caractéristiques liées au logement, à
l'hôte, à la localisation, aux prix, aux disponibilités et aux évaluations
des voyageurs.

Le jeu de données contient des variables de natures différentes : identifiants,
variables textuelles, variables catégorielles, coordonnées géographiques et
variables numériques.

On observe également dès cette première exploration la présence de valeurs
manquantes dans plusieurs variables. Leur traitement sera étudié lors des
étapes ultérieures du projet.

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15293 entries, 0 to 15292
Data columns (total 90 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            15293 non-null  int64  
 1   listing_url                                   15293 non-null  str    
 2   scrape_id                                     15293 non-null  int64  
 3   last_scraped                                  15293 non-null  str    
 4   source                                        15293 non-null  str    
 5   name                                          15293 non-null  str    
 6   description                                   14644 non-null  str    
 7   neighborhood_overview                         0 non-null      float64
 8   picture_url                                   15293 non-null  str    
 9   host_id                                       15293 non-null  int64  
 1

                            Structure et qualité initiale des données

Le jeu de données contient 15 293 observations et 90 variables.

Les variables sont réparties entre différents types : 18 variables entières,
42 variables flottantes et 30 variables textuelles.

L'analyse des valeurs non nulles montre une présence importante de données
manquantes dans plusieurs colonnes. Certaines variables sont même entièrement
vides, notamment `neighborhood_overview`, `host_since`, `host_response_time`,
`host_response_rate`, `host_acceptance_rate`, `host_thumbnail_url`,
`host_neighbourhood`, `host_total_listings_count`, `host_verifications`,
`neighbourhood`, `calendar_updated` et `instant_bookable`.

D'autres variables sont partiellement renseignées. Par exemple, `description`
est renseignée pour 14 644 annonces, tandis que `price` et
`price_quote_price_per_night` sont disponibles pour 13 355 observations.

Les variables relatives aux évaluations (`review_scores_*`) sont disponibles
pour environ 11 744 observations, ce qui indique que ces informations ne sont
pas présentes pour toutes les annonces.

Cette première analyse montre donc que la gestion des valeurs manquantes et
la sélection des variables constitueront des étapes importantes du
prétraitement. Cependant, aucune suppression ou imputation n'est effectuée
à ce stade : les décisions de nettoyage seront prises après une analyse plus
approfondie des données.

In [9]:
print(f"Nombre de variables avec au moins une valeur manquante : {len(missing)}")
print(f"Nombre de variables entièrement vides : {(df.isna().all()).sum()}")

Nombre de variables avec au moins une valeur manquante : 56
Nombre de variables entièrement vides : 12


In [10]:
# 1. One row represents:
print("rows:", f"{len(df):,}")
print("unique listing ids:", f"{df['id'].nunique():,}")
print("unique hosts:      ", f"{df['host_id'].nunique():,}")
print()
print("listings per host, top 5:")
print(df["host_id"].value_counts().head(5).to_string())

rows: 15,293
unique listing ids: 15,293
unique hosts:       4,595

listings per host, top 5:
host_id
346367515    588
1447144      448
21726991     359
32037490     293
4459553      243


Une ligne représente une annonce Airbnb. Le dataset contient 15 293 annonces et les 15 293 identifiants d'annonces sont uniques. En revanche, il n'y a que 4 595 hôtes, et certains hôtes possèdent plusieurs annonces (le plus grand en possède 588).

In [11]:
# 2. Number of completely unusable columns (a number):
audit = pd.DataFrame({
    "dtype": df.dtypes.astype(str),
    "missing_pct": (df.isna().mean() * 100).round(1),
    "n_unique": df.nunique(dropna=True),
})

audit.sort_values("missing_pct", ascending=False).head(20)

,dtype,missing_pct,n_unique
neighborhood_overview,float64,100.0,0
host_since,float64,100.0,0
host_response_time,float64,100.0,0
host_thumbnail_url,float64,100.0,0
host_acceptance_rate,float64,100.0,0
host_response_rate,float64,100.0,0
host_verifications,float64,100.0,0
neighbourhood,float64,100.0,0
host_total_listings_count,float64,100.0,0
host_neighbourhood,float64,100.0,0


In [12]:
empty_cols = [c for c in df.columns if df[c].isna().all()]
constant_cols = [c for c in df.columns if df[c].nunique(dropna=True) <= 1]

print(f"{len(empty_cols)} columns are 100% empty:")
for c in empty_cols:
    print(f"    {c}")

print(f"\n{len(constant_cols)} columns are constant: {constant_cols}")
print(f"\nUsable columns: {df.shape[1] - len(set(empty_cols) | set(constant_cols))} of {df.shape[1]}")

12 columns are 100% empty:
    neighborhood_overview
    host_since
    host_response_time
    host_response_rate
    host_acceptance_rate
    host_thumbnail_url
    host_neighbourhood
    host_total_listings_count
    host_verifications
    neighbourhood
    calendar_updated
    instant_bookable

13 columns are constant: ['scrape_id', 'neighborhood_overview', 'host_since', 'host_response_time', 'host_response_rate', 'host_acceptance_rate', 'host_thumbnail_url', 'host_neighbourhood', 'host_total_listings_count', 'host_verifications', 'neighbourhood', 'calendar_updated', 'instant_bookable']

Usable columns: 77 of 90


Le dataset contient 12 colonnes entièrement vides. Elles sont donc inutilisables car elles ne fournissent aucune information. Une colonne supplémentaire, scrape_id, est constante : elle contient une seule valeur pour toutes les observations et n'apporte donc pas d'information permettant de distinguer les annonces. Ainsi, 13 colonnes sont constantes ou entièrement vides, ce qui laisse 77 colonnes utilisables selon ce premier critère.

In [13]:
# Step 1: find the relevant column.
candidates = [c for c in df.columns if "instant" in c.lower()]
print("candidate columns:", candidates)

# Step 2: is it usable?
col = "instant_bookable"
print(f"\n{col}:")
print(f"  dtype        : {df[col].dtype}")
print(f"  missing      : {df[col].isna().mean() * 100:.1f}%")
print(f"  unique values: {df[col].nunique(dropna=True)}")
print(f"  value counts :\n{df[col].value_counts(dropna=False).to_string()}")

# Step 3: the reply.
print("""Reply to Client B ----------------- We cannot answer this question with the current data. The snapshot contains an `instant_bookable` field, but it is empty for all 15,293 listings - the 
platform stopped publishing it in this data source. Nothing can be inferred about an instant-booking premium from this 
dataset. Two options: (a) we obtain the field from your own booking system, where it is recorded per listing; or (b) we 
treat this as out of scope and focus on the pricing model, which the available fields do support. We recommend (b) for now
and can revisit (a) if you can export the field.""")

candidate columns: ['instant_bookable']

instant_bookable:
  dtype        : float64
  missing      : 100.0%
  unique values: 0
  value counts :
instant_bookable
NaN    15293
Reply to Client B ----------------- We cannot answer this question with the current data. The snapshot contains an `instant_bookable` field, but it is empty for all 15,293 listings - the 
platform stopped publishing it in this data source. Nothing can be inferred about an instant-booking premium from this 
dataset. Two options: (a) we obtain the field from your own booking system, where it is recorded per listing; or (b) we 
treat this as out of scope and focus on the pricing model, which the available fields do support. We recommend (b) for now
and can revisit (a) if you can export the field.


                                         A. Shape and grain
Le dataset contient 15 293 lignes et 90 colonnes. Une ligne représente une annonce Airbnb et les 15 293 identifiants 
d'annonces sont uniques. Cependant, les observations ne sont pas nécessairement indépendantes car 15 293 annonces 
appartiennent à seulement 4 595 hôtes, certains hôtes possédant plusieurs annonces. Le plus grand hôte possède 588 
annonces, ce qui doit être pris en compte lors de la séparation train/test et de la validation.

                                                B. Completeness

In [14]:
audit[audit["missing_pct"] < 100].sort_values(
    "missing_pct", ascending=False
).head(3)

,dtype,missing_pct,n_unique
host_about,str,31.8,2406
bathrooms,float64,23.3,21
review_scores_location,float64,23.2,134


Le dataset contient 12 colonnes entièrement vides, soit 100 % de valeurs manquantes. Parmi les colonnes qui sont 
partiellement renseignées, les trois plus incomplètes sont host_about avec 31,8 % de valeurs manquantes, bathrooms avec 
23,3 % et review_scores_location avec 23,2 %. Ces valeurs manquantes devront être prises en compte avant toute utilisation 
de ces variables dans un modèle.

                                                C. Types

In [15]:
df.dtypes.to_string()

'id                                                int64\nlisting_url                                         str\nscrape_id                                         int64\nlast_scraped                                        str\nsource                                              str\nname                                                str\ndescription                                         str\nneighborhood_overview                           float64\npicture_url                                         str\nhost_id                                           int64\nhost_url                                            str\nhost_profile_id                                 float64\nhost_profile_url                                    str\nhost_name                                           str\nhost_since                                      float64\nhosts_time_as_user_years                        float64\nhosts_time_as_user_months                       float64\nhosts_time_as_host_years      

Deux types de variables nécessitent une attention particulière. La colonne price est stockée comme une chaîne de caractères
(str) alors qu'elle représente une valeur numérique correspondant au prix d'une annonce; elle devra donc être nettoyée et 
convertie avant une analyse numérique. De même, plusieurs variables comme first_review, last_review et last_scraped sont 
stockées comme str alors qu'elles représentent des dates. Elles devraient être converties dans un format de date afin de 
pouvoir effectuer correctement des opérations temporelles.

                                                 D. Questions
Les données permettent d'étudier plusieurs questions:
1. Comment le prix des annonces varie selon le type de logement?
2. Comment les caractéristiques et les prix des annonces varient-ils selon les quartiers de Barcelone ?
3. Comment la disponibilité des logements varie selon leur type? 

En revanche, les données ne permettent pas de déterminer si les annonces avec réservation instantanée ont un prix plus 
élevé, car la variable instant_bookable est entièrement vide dans ce snapshot. Cette question nécessiterait des données 
complètes sur la réservation instantanée et un prix exploitable.

                                              E. Problem statement
 CLIENT A:
Pour la Direction du Tourisme de Barcelone, l'unité d'observation est une annonce Airbnb. L'objectif est 
d'identifier les annonces susceptibles de fonctionner sans licence touristique valide afin d'aider les services 
d'inspection à cibler leurs contrôles. La cible serait donc le statut de conformité de l'annonce. Le succès serait une 
sélection d'annonces suffisamment fiable et défendable pour permettre aux services concernés de concentrer leurs 
ressources d'inspection, plutôt que de simplement maximiser un score de performance du modèle.

 CLIENT B:
Pour la société de gestion immobilière, l'unité d'observation est une annonce ou un appartement. L'objectif 
est d'estimer le prix auquel une nouvelle annonce pourrait être proposée à partir de ses caractéristiques. La cible 
serait donc le prix de l'annonce. Le succès serait d'obtenir une estimation suffisamment proche des prix pertinents du 
marché pour aider l'entreprise à fixer un prix de mise en location, et non simplement d'obtenir une bonne performance sur 
une métrique de régression.

La classification est une méthode d’apprentissage automatique qui consiste à prédire une catégorie ou une classe à partir de données d’entrée, tandis que la régression consiste à prédire une valeur numérique continue.
Ainsi, 
Pour le CLIENT A, il s'agit d'un problème de classification. Ici, l'unité d'observation est une annonce Airbnb et la cible correspond au statut de conformité de l'annonce par rapport à sa licence touristique. Le résultat utile pour le client serait une sélection d'annonces à cibler pour les inspections, accompagnée d'une estimation suffisamment fiable pour justifier ces choix.
Pour le CLIENT B, il s'agit d'un problème de régression. Ici, l'unité d'observation est une annonce ou un appartement et la cible est son prix. Le résultat utile pour le client serait une estimation de prix suffisamment proche des prix pertinents du marché pour aider à fixer le prix d'une nouvelle annonce.

Plusieurs erreurs doivent être évitées. Il ne faut pas considérer toutes les lignes comme indépendantes, car plusieurs 
annonces peuvent appartenir au même hôte. Il ne faut pas non plus se fier uniquement au nom d'une colonne : par exemple, 
`instant_bookable` semble pertinente pour la question du Client B mais elle est entièrement vide. De même, l'observation 
de quelques lignes avec `head()` ne permet pas d'évaluer la qualité globale des données. Enfin, le succès d'un projet ne 
doit pas être défini uniquement par une métrique et il est préférable de comprendre et préparer les données avant de  commencer la modélisation.

Une colonne peut être inutilisable même si elle ne contient aucune valeur manquante. Elle peut être constante, avoir un mauvais type, contenir des valeurs incohérentes ou aberrantes, ou encore être redondante avec une autre variable. Par 
exemple, `scrape_id` n'a pas de valeurs manquantes mais elle est constante, donc elle n'apporte aucune information pour distinguer les annonces.

Les données ne sont pas neutres : elles proviennent d'un projet de collecte portant sur les locations Airbnb et n'ont pas été recueillies spécifiquement pour répondre aux besoins de nos deux clients. Elles reflètent donc les informations rendues publiques sur la plateforme ainsi que les choix et les limites du processus de collecte. Deux clients peuvent ainsi utiliser le même dataset pour des objectifs différents et parvenir à des conclusions différentes selon leurs intérêts et leurs questions.


                            Milestone 1: Cleaning pipeline and decision log 

Ce Milestone 1 a pour objectif de construire un pipeline de nettoyage des données reproductible et documenté pour le projet longitudinal sur le marché de la location de courte durée à Barcelone.

Avant de nettoyer les données, nous devons d'abord mettre en place une séparation correcte entre les données d'entraînement et les données de test. Cette étape est particulièrement importante dans ce jeu de données, car plusieurs annonces peuvent appartenir au même hôte. Une séparation aléatoire classique pourrait donc placer des annonces d'un même hôte dans les deux ensembles et introduire une fuite d'information entre le train et le test. Nous utiliserons donc une séparation par groupe, basée sur host_id.

Une fois le jeu de test scellé, toutes les décisions de nettoyage seront prises uniquement à partir des données d'entraînement. Nous identifierons ensuite les principaux problèmes de qualité présents dans les données : types incorrects, valeurs structurées dans une seule cellule, valeurs impossibles ou extrêmes, valeurs manquantes et doublons potentiellement trompeurs.

L'objectif n'est pas de supprimer automatiquement toutes les valeurs atypiques ou manquantes. Chaque transformation devra être justifiée en précisant pourquoi elle est réalisée, ce qui aurait été perdu sans cette transformation et quel biais elle pourrait éventuellement introduire.

Le résultat attendu est donc un pipeline de nettoyage reproductible, accompagné d'un journal des décisions permettant de retracer et de justifier chaque choix effectué.

                            1. Chargement et préparation de l'environnement

Nous commençons par charger le jeu de données brut ainsi que les paramètres utilisés pendant le projet. La graine aléatoire est fixée afin de garantir la reproductibilité des opérations qui impliquent une part d'aléatoire.

À ce stade, aucune transformation n'est appliquée aux données. Le jeu de données est conservé dans son état brut afin de pouvoir effectuer la séparation train/test avant toute décision de nettoyage.

In [1]:
import sys
from pathlib import Path

# Same root-finding walk as Session 1: make src/ importable wherever we launched.
ROOT = Path.cwd()
while not (ROOT / "src").is_dir() and ROOT != ROOT.parent:
    ROOT = ROOT.parent
sys.path.insert(0, str(ROOT))

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

# split_by_host is the grouped split we are about to justify from first principles.
from src.data import PATHS, SEED, load_raw, set_seed, split_by_host

# One consistent visual style for every chart in the session, set once.
sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", 50)
pd.set_option("display.width", 140)
# Seed before anything random happens, including the split below.
set_seed(SEED)

df = load_raw()
print(f"loaded {df.shape[0]:,} rows x {df.shape[1]} columns")

loaded 15,293 rows x 90 columns


Le jeu de données contient 15 293 observations et 90 variables. Chaque observation correspond à une annonce Airbnb. Le chargement est effectué à partir des données brutes fournies dans le projet, sans modification préalable.